src/mesh/domain.hpp

```diff
@@ -0,0 +1,181 @@
+      : IndexShape( std::vector<int> {nx1,nx2,nx3}, ng) {};
+
+      IndexShape(const int & nx1, const int & nx2, const int & ng) 
+      : IndexShape( std::vector<int> {nx1,nx2}, ng) {};
+
+      IndexShape(const int & nx1, const int & ng) 
+      : IndexShape( std::vector<int> {nx1}, ng) {};
+
+      IndexShape(const std::vector<int> & interior_dims, const int & ng) { 
+        assert(interior_dims.size()<=NDIM && "IndexShape cannot be initialized, the number of "
+              "dimensions exceeds the statically set dimensions, you will need to change the NDIM "
+              "constant.");
+        for( int dim=1, index=0; dim<=NDIM; ++dim, ++index){
+          if (dim <= interior_dims.size()) {
+            assert( interior_dims.at(index) > 0 && "IndexShape cannot be initialized with fewer "
+                "than 1 interior cells for each dimension");
+            x_[index].start = ng;
+            x_[index].end = x_[index].start + interior_dims.at(index) - 1;
+            entire_ncells_[index] = interior_dims.at(index) + 2*ng;
+          } else {
+            x_[index].start = 0;
+            x_[index].end = 0;
+            entire_ncells_[index] = 1;
+          }
+        }
+      };
+
+      std::array<IndexRange,NDIM> GetBounds(const IndexShapeType & type) const noexcept {
+        if( type==IndexShapeType::entire ){
+          std::array<IndexRange,NDIM> bounds;
+          for (int index=0; index<NDIM;  ++index ){
+            bounds[index].start = 0;
+            bounds[index].end = entire_ncells_[index];
+          }
+          return bounds;
+        }
+        return x_; 
+      }
+      
+      inline int is(const IndexShapeType & type) const 
+      { return (type==IndexShapeType::entire) ? 0 : x_[0].start; }
+      
+      inline int js(const IndexShapeType & type) const
+      { return (type==IndexShapeType::entire) ? 0 : x_[1].start; }
+      
+      inline int ks(const IndexShapeType & type) const
+      { return (type==IndexShapeType::entire) ? 0 : x_[2].start; }
+      
+      inline int ie(const IndexShapeType & type) const 
+      { return (type==IndexShapeType::entire) ? entire_ncells_[0]-1 : x_[0].end; }
+      
+      inline int je(const IndexShapeType & type) const 
+      { return (type==IndexShapeType::entire) ? entire_ncells_[1]-1 : x_[1].end; }
+      
+      inline int ke(const IndexShapeType & type) const 
+      { return (type==IndexShapeType::entire) ? entire_ncells_[2]-1 : x_[2].end; }
+
+      inline int nx1(const IndexShapeType & type) const 
+      { return (type==IndexShapeType::entire) ? entire_ncells_[0] : x_[0].ncells(); }
+      
+      inline int nx2(const IndexShapeType & type) const 
+      { return (type==IndexShapeType::entire) ? entire_ncells_[1] : x_[1].ncells(); }
+      
+      inline int nx3(const IndexShapeType & type) const 
+      { return (type==IndexShapeType::entire) ? entire_ncells_[2] : x_[2].ncells(); }
```

Comment:
My suggestion is to do the following:

```cxx
  struct IndexRange {
    IndexRange() {}
    IndexRange(int start, int end) :
      s(start), e(end) {
      assert(e >= s);
    }

    int s = 0; /// Starting Index (inclusive)
    int e = 0; /// Ending Index (inclusive)

    int ncells() const noexcept { 
      assert(e >= s);
      return e - s + 1; 
    }
  };
```

And in `IndexShape`:

```cxx
inline auto InteriorXRange() const { return x_[0]; }
inline auto InteriorYRange() const { return x_[1]; }
inline auto InteriorZRange() const { return x_[2]; }

inline auto EntireXRange() const { return IndexRange(0, entire_ncells[0] - 1); }
inline auto EntireYRange() const { return IndexRange(0, entire_ncells[1] - 1); }
inline auto EntireZRange() const { return IndexRange(0, entire_ncells[2] - 1); }

inline auto InteriorNX() const { return x_[0].ncells(); }
inline auto InteriorNY() const { return x_[1].ncells(); }
inline auto InteriorNZ() const { return x_[2].ncells(); }

inline auto EntireNX() const { return entire_ncells[0]; }
inline auto EntireNY() const { return entire_ncells[1]; }
inline auto EntireNZ() const { return entire_ncells[2]; }
```
Maybe we'd need some asserts for `NDIM`.

Then one would use this as:
```cxx
auto i = pmb->cellbounds.InteriorXRange(); // or EntireXRange()
auto j = pmb->cellbounds.InteriorYRange();
auto k = pmb->cellbounds.InteriorZRange();

// now use i.s, i.e, j.s, j.e, k.s, k.e instead of is, ie, js, je, ks, ke
```

Reply:
I think I would prefer to keep the syntax consistent

    pmb->cellbounds.InteriorIRange();
    pmb->cellbounds.InteriorJRange();
    pmb->cellbounds.InteriorKRange();

Reply:
@jlippuner the problem with having the Interior, and Entire in the method name is it does not allow flexibility at runtime when there may be cases when one or the other are being used. E.g. in vtk.cpp

    IndexShapeType shape_type = interior;
    if (output_params.include_ghost_zones) {
      shape_type = entire;
    } 

I therefore would go with the second option of a scoped enum.

Reply:
I lean towards following @jmstone here... in a physics code, I think `is`,`ie`, etc., are fairly consistent. That said, `ib.s` is pretty close to this, and I think it's unambiguous.

@jmstone what do you think of `ib.s` and `ib.e`?

Reply:
I think ib.s and ib.e would work.  They are close enough and I agree they are unambiguous.  The shape type 'interior' will still need careful documentation, however, since that can be interpreted differently for different variable centerings.  For cell-centered it is obviously [is:ie,js:je,ks:ke], but for face-centered is it the same or [is:ie+1,js:je,ks:ke]?? (for x1-faces).